src/mem/freelist.h

```diff
@@ -283,63 +251,18 @@ namespace snmalloc
       return true;
     }
 
-    bool debug_different_slab(CapPtr<FreeObject, CBAlloc> n)
-    {
-      for (size_t i = 0; i < LENGTH; i++)
-      {
-        if (!different_slab(address_cast(end[i]), n))
-          return false;
-      }
-      return true;
-    }
-
     /**
      * Adds an element to the builder
      */
     void add(CapPtr<FreeObject, CBAlloc> n, LocalEntropy& entropy)
     {
-      SNMALLOC_ASSERT(!debug_different_slab(n) || empty());
-
       uint32_t index;
       if constexpr (RANDOM)
         index = entropy.next_bit();
       else
         index = 0;
 
-      end[index]->store(n, get_prev(index), entropy);
-#ifdef CHECK_CLIENT
-      prev[index] = get_curr(index);
-#endif
-      end[index] = &(n->next_object);
-    }
-
-    /**
-     *  Calculates the length of the queue.
-     *  This is O(n) as it walks the queue.
-     *  If this is needed in a non-debug setting then
-     *  we should look at redesigning the queue.
-     */
-    size_t debug_length(LocalEntropy& entropy)
-    {
-      size_t count = 0;
-      for (size_t i = 0; i < LENGTH; i++)
-      {
-        uint32_t local_prev = HEAD_KEY;
-        EncodeFreeObjectReference* iter = &head[i];
-        CapPtr<FreeObject, CBAlloc> prev_obj = iter->read(local_prev, entropy);
-        UNUSED(prev_obj);
-        uint32_t local_curr = address_cast(&head[i]) & 0xffff'ffff;
-        while (end[i] != iter)
-        {
-          CapPtr<FreeObject, CBAlloc> next = iter->read(local_prev, entropy);
-          check_client(!different_slab(next, prev_obj), "Heap corruption");
-          local_prev = local_curr;
-          local_curr = address_cast(next) & 0xffff'ffff;
-          count++;
-          iter = &next->next_object;
-        }
-      }
-      return count;
+      end[index] = FreeObject::store(end[index], n, entropy);
     }
 
     /**
```

Comment:
Annoyingly, I can't comment on lines not considered part of the context diff.  In any case, just below this text field, the comment says "Termination uses the bottom bit" which I think is no longer true?

Reply:
Thanks, good catch